rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,383 @@
+    part_size_multiple: 1,
+  },
+  {
+    type: 'multipart',
+    max_size: 5 * 1024 * 1024 * 1024,
+    min_size: 0,
+    max_part_count: 10000,
+    min_part_count: 1,
+    min_part_size: 5 * 1024 * 1024,
+    max_part_size: 5 * 1024 * 1024,
+    part_size_multiple: 1,
+  },
+]
+```
+
+All services will use a shape like this, changing their values to reflect their
+reality.  This format will give the uploader the ability to choose from the
+supported types of uploads.  Only those upload types which work with this
+upload flow will be supported, regardless of whether more options are available
+in the underlying API.
+
+The keys for this response will have the following meaning:
+
+* `type`: an identifier for the style of upload, unique to each underlying
+  service
+* `min`/`max_size`: The size range for the overall object in bytes
+* `min`/`max_part_count`: The range of number of parts allowed
+* `min`/`max_part_size`: The size range for each part in bytes
+* `part_size_multiple`: If a service requires parts be a multiple of a number,
+  this is the value.
+
+The ranges must all work together such that the most strict condition is not
+exceeded.  For example if the maximum number of parts is 10,000 and you want to
+use 5MB parts, you are limited to 5MB * 10,000, not 5TB.
+
+Of note is that the minimum part size is interpreted slightly differently.  In
+order to support last part part sizes which are not the minimum part size, the
+last part may be smaller than the minimum value specified here.
+
+This endpoint provides advisory information and does not obviate the service from
+checking these values to be valid when the object is created.  The object creation
+endpoint must always validate its provided values.
+
+The value of this URL is cachable for at least 30 minutes, which implies that
+origin to backend mapping must be stable for 30 minutes at a time.
+
+### Deletion
+
+```
+DELETE /objects/:name
+DELETE /caches/:id/objects/:name
+```
+
+This service has at least one copy of each stored file.  Any copies stored in
+addition to this original and authoritative copy are considered to be cached
+copies.
+
+The first endpoint will delete the original copy of the file as well as
+initiating the deletion of any cached copies.  On success, it will return a 200
+to reflect the success of deleting the original copy.
+
+The second endpoint would attempt delete all of the cached copies in the
+various locations, leaving the original file intact.  The cached copies are
+those copies which are created in other regions or services to provide more
+local copies of each object.  This cache purging must return a `202 Accepted`
+response and not `200/204` due to the nature of how caches work in the real
+world.
+
+The `id` parameter in the second endpoint would specify a specific cache id for
+which the deletion is requested.  A special reserved value `all` would request
+that all non-canonical copies of the object be deleted.
```

Comment:
Does this `id` appear anywhere else?  How would a user get this value?  Perhaps a `GET /caches` to list all caches, or `GET /caches/objects/:name` to get the caches for a specific object?  Or perhaps downloads can come with an X-Objservice-Cache-Id header on the initial redirect?